File: python/src/github_wiki_organiser/unknown_wiki_list_exporter_for_llm.py
import os
from typing import Any

from .application import Application

DEFAULT_OUTPUT_FILENAME = 'unknown_wiki_list_for_llm.txt'


class UnknownWikiListExporterForLLM(Application):
    """Writes a flat list of the wikis under the configured "unknown
    owner/category" namespace in a form suitable for feeding to an LLM."""

    _cached_unknown: list[str]

    def __init__(self, output: str | None = None, **options: Any) -> None:
        super().__init__(**options)
        self._path_to_export = os.path.join(
            self._base_path, output or DEFAULT_OUTPUT_FILENAME)

    # private

    def _run(self) -> str:
        wikis = self._unknown_wiki_list_for_llm()
        with open(self._path_to_export, 'wb') as f:
            f.write(('\n'.join(wikis) + '\n').encode())
        return self._path_to_export

    def _target_namespace(self) -> str:
        return self._config.llm_target_namespace(self._group_by,
                                                 self._language)

    def _unknown_wiki_list_for_llm(self) -> list[str]:
        if hasattr(self, '_cached_unknown'):
            return self._cached_unknown
        plain = self._plain_wiki_maps()
        target = self._target_namespace()
        self._cached_unknown = list(plain.get(target, []))
        return self._cached_unknown

File: RubyGem/lib/spreen/wiki/unknown_wiki_list_exporter_for_llm.rb
# frozen_string_literal: true
# rbs_inline: enabled

require_relative 'application'

module Spreen
  module Wiki
    # Writes a flat list of the wikis under the configured "unknown
    # owner/category" namespace in a form suitable for feeding to an LLM.
    class UnknownWikiListExporterForLLM < Application
      DEFAULT_OUTPUT_FILENAME = 'unknown_wiki_list_for_llm.txt'

      # @rbs base_path: String
      # @rbs group_by: String
      # @rbs language: String
      # @rbs home_overflow: (bool | String)
      # @rbs output: String?
      # @rbs **options: untyped
      # @rbs return: void
      def initialize(base_path: Dir.pwd, group_by: 'Owner', language: 'English', home_overflow: 'false',
                     output: nil, **)
        super(base_path:, group_by:, language:, home_overflow:, **)
        @path_to_export = resolve_output_path(output || DEFAULT_OUTPUT_FILENAME)
      end

      # @rbs return: String
      def run
        File.open(path_to_export, 'wb') { |f| f.puts(unknown_wiki_list_for_llm) }
        path_to_export
      end

      private

      attr_reader :path_to_export #: String

      # @rbs filename: String
      # @rbs return: String
      def resolve_output_path(filename)
        File.absolute_path?(filename) ? filename : File.join(base_path, filename)
      end

      # @rbs return: String
      def target_namespace
        config.llm_target_namespace(group_by, language)
      end

      # @rbs return: Array[String]
      def unknown_wiki_list_for_llm
        @unknown_wiki_list_for_llm ||= plain_wiki_maps.slice(target_namespace).values.flatten
      end
    end
  end
end
